suggested documentation of var-type return values

capocasa opened the pull request:
Discovered undocumented behavior in forum discussion, suggested documentation contained in pull request.

Comments and requests for changes welcome.

capocasa committed:
suggested documentation of var-type return values

c-blake commented:
You may also want to mention that you can preserve ability to edit with `addr`, as in
```Nim
var someVal = addr someTable.mgetOrPut(key, initial)
# Now changing someVal changes the memory in someTable
```
Or look at other ways to describe this in https://forum.nim-lang.org/t/9856

juancarlospaco commented on doc/manual/var_t_return.md:
```diff
@@ -22,3 +22,13 @@ then it has to be derived from the routine's first parameter:
 lifetime of the first parameter and that is enough knowledge to verify
 memory safety at the call site.
+
+When a `var`-typed return value is assigned to a local variable using a
```
What happens when you assign to a global `var` in the global root namespace?.

I would mention or link `out` parameters too.

capocasa replied:
local was a misnomer, changed to `var`-declared value

capocasa commented:
I have modified the documentation to include `addr`, replace the incorrect word `local`, added an example and moved most of it to the `var`-statement documentation, which seems more natural. Leaving out `var`-params, I had no trouble learning about them from existing docs.

More feedback quite welcome if you find something you think needs changing.

c-blake commented on doc/manual.md:
```diff
@@ -3060,6 +3060,30 @@ But these will compile successfully:
+x = "that"               # changes the copy
+echo foo.x               # prints "this", the original value is unchanged
+
+var z = addr foo.getX    # puts a reference to foo.x into x
```
In the comment I think you meant to say "into z"

beef331 replied:
I'd also not say "a reference" as it's not a `ref` but a `ptr` so  something along "stores pointer to foo.x" is more clear.

juancarlospaco commented on doc/manual.md:
````diff
@@ -3060,6 +3060,30 @@ But these will compile successfully:
+```nim
+var foo = Foo(x: "this")
+
+proc getX(f: var Foo): var string
````
This is not valid code ?, `=` missing.

I would add a `type Foo` for completeness.

capocasa replied:
Fixed! Thanks

c-blake commented:
I'm not sure if this is the place for an example showing it, but for compound (object/tuple) return types some features compose to let you say this (without the explicit `[]` derefer after `z`):
```Nim
type Foo = tuple[x: string]
proc getX(f: var Foo): var Foo = f
var foo = ("1",)    #  ^^^ object return
var z = addr foo.getX
echo foo
z.x = "2"           # no []
echo foo
```
but `z[].x` also works.  Mostly just an FYI related to the long Forum thread.

c-blake commented:
Similarly to my last comment, another option for "leaf types" is:
```Nim
from std/decls import byaddr
type Foo = tuple[x: string]
proc getX(f: var Foo): var string = f.x
var foo = ("1",)
var z {.byaddr.} = foo.getX # `let` also works!
echo foo
z = "2"           # another no [] needed
echo foo
```
Both of these complement the "write a proc receiving the var" approach.  Nim is choice. ;-)

capocasa commented:
I originally excluded byaddr because I thought it was experimental and compiler-flag-enabled but since it's not, I added it!

Also fixed the code example and added one for byaddr.

Since addr returns a pointer I think I'll leave it to the linked pointer-and-reference-types documentation to explain all the ways you can dereference, I didn't have any trouble learning those from the manual.

I have a feeling the addition is pretty good now, thanks for all the feedback!

c-blake commented on doc/manual.md:
```diff
@@ -3060,6 +3060,40 @@ But these will compile successfully:
+echo foo.x                   # prints "this", the original value is unchanged
+
+import std/decls
+var y {.byaddr.} = foo.getX  # put the original foo.x into x
```
Now you have 2 copy-pasta mistakes in comments - this one where "into x" should be "into y" ...

capocasa replied:
Whoops. Fixed!

c-blake commented on doc/manual.md:
```diff
@@ -3060,6 +3060,40 @@ But these will compile successfully:
+y = "that"                   # change the original value
+echo foo.x                   # prints "that", the original value was changed
+
+var z = addr foo.getX        # puts a reference to foo.x into x
```
... and this one where "into x" should be "into z".

capocasa replied:
Fixed too

c-blake commented:
FWIW, almost all of `doc/manual.md` is <= 80 ASCII text columns.  So, it might be nice to wrap these lines at 80 for this section, too. Personally, I do not read doc/* rendered as HTML and I am probably not the only one and these kinds of updates often work well as "combination improvements".  { And no, this is not an "April fool's" comment. ;-) }

capocasa commented:
No problem! Will do- thanks for taking the time to bring me up to speed

capocasa commented:
Unfortunately, we're going to have to wait until the custom AI model I use instead of Vim these days is trained to wrap at 80 characters. I have a team of workers on it as we speak.

c-blake commented:
Of course, I realize you were joking (and it was great). More seriously, it did inspire me to document & include this [program I wrote a while back](https://github.com/c-blake/bu/blob/main/doc/widths.md).

capocasa commented:
It took a few overtime shifts but they got it done! Nice tool

capocasa commented:
If there are no further comments, would somone kindly put their standing in the community on the line for me ;) and merge this?

Araq commented:
> If there are no further comments, would somone kindly put their standing in the community on the line for me ;) and merge this?

I'm still reviewing it. Tbh I'm not too happy with it so I might rewrite the section. What is really going on here is that type modifiers are "skipped" in the type inference for `var x = f()`. This then has the effects that you describe, but your wording makes it sound like there are lots of special cases going on when in reality there is nothing more to it, modifiers are skipped.

capocasa commented:
I see! It's always a challenge to convey internal mechanisms enough but not so much it distracts from what you can do with them.

I'd be happy to re-write as well with guidance- if you would, can you give a link where I can learn more about type modifiers? Feel free to do it yourself if you prefer of course, I'm fine either way.

Araq commented:
> can you give a link where I can learn more about type modifiers?

That's the issue, they are criminally under-documented. Type modifiers in Nim are `lent`, `var`, `out` and `sink`.

capocasa commented:
Then we'd better fix that before Europol catches wind, I hear conditions in the underdocumentor's prison can get a bit crowded.

> Type modifiers in Nim are lent, var, out and sink.

Okay now I understand, I thought 'type modifier' was some kind of internal implementation detail.

I'll see if I can make the conceptual simplicity clearer, although I'd suggest to consider it might be more important for new users to just explain what it does. It's the paradox of the meta-level- whoever has to live with the implications needs to know the details of what they are.
